File: src/main/java/ar/com/flow/bankaccount/transaction/transfer/Transfer.java
package ar.com.flow.bankaccount.transaction.transfer;

import ar.com.flow.bankaccount.BankAccount;
import ar.com.flow.bankaccount.transaction.Action;
import ar.com.flow.bankaccount.transaction.Step;
import ar.com.flow.bankaccount.transaction.receipt.CreditStamper;
import ar.com.flow.bankaccount.transaction.receipt.DebitStamper;
import ar.com.flow.bankaccount.transaction.Transaction;
import ar.com.flow.bankaccount.transaction.withdrawal.SufficientFunds;
import ar.com.flow.money.Money;
import lombok.RequiredArgsConstructor;

public class Transfer {
    public static BankTransferBuilder from(BankAccount debitAccount) {
        return new BankTransferBuilder(debitAccount);
    }

    @RequiredArgsConstructor
    public static class BankTransferBuilder {
        private final BankAccount debitAccount;
        private BankAccount creditAccount;

        public BankTransferBuilder to(BankAccount creditAccount) {
            this.creditAccount = creditAccount;
            return this;
        }

        public Transaction amount(Money amountToTransfer) {
            return Transaction.builder()
                    .precondition(new SufficientFunds(debitAccount, amountToTransfer))
                    .precondition(new DifferentAccounts(debitAccount, creditAccount))
                    .step(new Step(debitAccount, Transfer.debitReceipt(creditAccount)))
                    .step(new Step(creditAccount, Transfer.creditReceipt(creditAccount)))
                    .amount(amountToTransfer)
                    .build();
        }
    }

    public static DebitStamper debitReceipt(BankAccount account) {
        return new DebitStamper(account, Action.Transfer);
    }

    public static CreditStamper creditReceipt(BankAccount account) {
        return new CreditStamper(account, Action.Transfer);
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/domain/Transfer.kt
package ar.com.flow.bankaccount.domain

import ar.com.flow.bankaccount.domain.transaction.Builder
import ar.com.flow.bankaccount.domain.transaction.Transaction
import ar.com.flow.bankaccount.domain.transaction.receipt.*
import ar.com.flow.bankaccount.domain.transaction.steps.Step
import ar.com.flow.bankaccount.domain.withdrawal.SufficientFunds
import ar.com.flow.money.Money

internal object Transfer {
    fun from(debitAccount: BankAccount): BankTransferBuilder {
        return BankTransferBuilder(debitAccount)
    }

    internal class BankTransferBuilder(private val debitAccount: BankAccount) {
        private lateinit var creditAccount: BankAccount

        fun to(creditAccount: BankAccount): BankTransferBuilder {
            this.creditAccount = creditAccount
            return this
        }

        fun amount(amountToTransfer: Money): Transaction {
            return Builder()
                .precondition(SufficientFunds(debitAccount, amountToTransfer))
                .precondition(AccountAreDifferent(debitAccount, creditAccount))
                .step(Step(debitAccount, TransferDebitPrint(creditAccount), TransferDebitScratch(creditAccount)))
                .step(Step(creditAccount, TransferCreditPrint(creditAccount), TransferCreditScratch(creditAccount)))
                .amount(amountToTransfer)
                .build()
        }
    }
}